### PR TITLE
fix(tests): remove terminal from SANDBOX_ALLOWED_TOOLS and update tests

### PR DESCRIPTION
**Pull request by LifeJiggy**
Removes `terminal` from `SANDBOX_ALLOWED_TOOLS` (sandbox should use sandbox-native tools, not terminal) and updates 5 tests to use `web_search`/`write_file` instead.

**Changes:**
- Removed `terminal` from `SANDBOX_ALLOWED_TOOLS` in `tools/code_execution_tool.py`
- Updated `test_generates_subset`, `test_non_allowed_tools_ignored`, `test_none_enabled_tools_uses_all`, `test_empty_enabled_tools_uses_all`, `test_nonoverlapping_tools_fallback`

Part of #24942

### CODE REVIEW

**Comment by LifeJiggy** [comment]
@egilewski pTAL

**Review by egilewski — CHANGES_REQUESTED** [review]
Needs rework.

I validated PR head `9286acd3321f7c4b3125046b91754fc22e5059fe` against current GitHub `main` `57775e9e161087dbe55e096c038f512233c03381`. Removing `terminal` from `SANDBOX_ALLOWED_TOOLS` is the intended security direction, but the test migration is incomplete: several remaining execute-code tests still import `terminal` from the generated sandbox module, so the focused suite fails after the allowlist change.

Evidence:
- Direct probe on the PR confirms `terminal_allowed=False` and the generated `hermes_tools` module no longer contains `def terminal(`.
- Current main baseline for the affected tests passed: `python -B -m pytest -o addopts='' -p no:cacheprovider tests/tools/test_code_execution.py::TestExecuteCode::test_single_tool_call tests/tools/test_code_execution.py::TestExecuteCode::test_multi_tool_chain tests/tools/test_code_execution.py::TestExecuteCode::test_concurrent_tool_calls_match_responses -q` -> `3 passed`.
- Current-main replay of this tip commit failed the same command with `3 failed`; all failures are `ImportError: cannot import name 'terminal' from 'hermes_tools'`.
- Full PR-head focused run with a short `TMPDIR` to avoid Unix socket path limits failed `tests/tools/test_code_execution.py -q` with `3 failed, 64 passed`; the failing tests are the same remaining terminal-import cases.
- `python -B -m compileall -q tools/code_execution_tool.py tests/tools/test_code_execution.py` and `git diff --check upstream/main..upstream/pr/42752` passed.
- CodeRabbit on the current-main replay diff returned `No findings`.

Please update the remaining execute-code tests to use allowed sandbox tools, or explicitly move terminal-import behavior into a negative test that expects the import/call to fail.

Signed: GPT-5.5-xhigh in Codex

**Comment by LifeJiggy** [comment]
@egilewski Fixed! Found and migrated the 4 remaining tests that still imported 	erminal from hermes_tools:

- 	est_single_tool_call - now uses web_search
- 	est_multi_tool_chain - now uses web_search + 
ead_file
- 	est_concurrent_tool_calls_match_responses - now uses web_search with updated mockn- 	est_excluded_tool_returns_error - now uses ision_analyze (excluded tool) instead of 	erminal

Full test suite passes: 36 passed, 31 skipped (UDS tests skip on Windows).

**Comment by LifeJiggy** [comment]
@egilewski Fixed! Found 3 more generate_hermes_tools_module calls in 	est_code_execution_windows_env.py that still used 	erminal as the tool name. Updated them to use web_search instead.

Full test suite passes: 109 passed, 43 skipped across 	est_code_execution.py, 	est_code_execution_windows_env.py, 	est_code_execution_modes.py, and 	est_execute_code_approval_cluster.py.

**Review by egilewski — CHANGES_REQUESTED** [review]
Needs rework.

I re-reviewed PR head `a7fa68a3bed60b3b2878091cf7ed57dde9048ba8` against current GitHub `main` `d046169646b03d889bfadac3329970559b655a27`. The original `tests/tools/test_code_execution.py` failures are fixed, but there are still positive-path mode tests that import `terminal` from `hermes_tools` after the PR removes `terminal` from `SANDBOX_ALLOWED_TOOLS`.

Evidence:
- Current main baseline: `python -B -m pytest -o addopts='' -p no:cacheprovider tests/tools/test_code_execution_modes.py::TestExecuteCodeModeIntegration::test_project_mode_can_still_import_hermes_tools tests/tools/test_code_execution_modes.py::TestExecuteCodeModeIntegration::test_strict_mode_can_still_import_hermes_tools -q` passed (`2 passed`).
- Current-main replay of this PR failed the same command (`2 failed`). Both failures still come from `tests/tools/test_code_execution_modes.py`, where the test bodies import `terminal` from `hermes_tools` and expect success.
- `python -B -m compileall -q tools/code_execution_tool.py tests/tools/test_code_execution.py tests/tools/test_code_execution_windows_env.py tests/tools/test_code_execution_modes.py` passed.
- `git diff --check` passed.
- CodeRabbit was not run for this re-review because local validation already found a request-changes blocker.

Please migrate those mode tests to allowed sandbox tools as well, or change them into explicit negative tests if they are intended to cover unavailable-tool behavior.

Signed: GPT-5.5-xhigh in Codex

**Comment by LifeJiggy** [comment]
@egilewski Fixed! Migrated the 2 remaining 	erminal imports in 	est_code_execution_modes.py:

- 	est_project_mode_can_still_import_hermes_tools - now uses web_search
- 	est_strict_mode_can_still_import_hermes_tools - now uses web_search

Full suite: 109 passed, 43 skipped. All 	erminal references from sandbox-allowed-tools tests should now be gone.

**Comment by egilewski** [comment]
Tests still fail though, is that GitHub Actions acting up?

**Comment by LifeJiggy** [comment]
@egilewski Fixed the root cause! Downloaded the CI logs and found the actual failures weren't ImportError but mock mismatches:

**Slice 4** (	est_code_execution.py):
- 	est_single_tool_call - mock returned 
esults dict instead of output for web_search`n- 	est_multi_tool_chain - same mock format issue

**Slice 6** (	est_code_execution_modes.py):
- 	est_project_mode_can_still_import_hermes_tools - mock only handled 	erminal, not web_search`n- 	est_strict_mode_can_still_import_hermes_tools - same

**Fix:** Updated both mock dispatchers to handle web_search with the same output format as 	erminal.

Full suite: 60 passed, 43 skipped.

**Comment by LifeJiggy** [comment]
@egilewski Fixed! Root cause was mock dispatchers not handling web_search:
- 	est_code_execution.py: mock returned 
esults dict instead of output`n- 	est_code_execution_modes.py: mock only handled 	erminal, not web_search

Updated both mocks. 60 passed, 43 skipped.

**Comment by egilewski** [comment]
Some tests still fail.

**Comment by LifeJiggy** [comment]
Some test still fail and from my side the issue is fixed and passed locally

**Comment by LifeJiggy** [comment]
Note
The test (3) CI failure is a pre-existing flaky timeout in tests/tools/test_browser_homebrew_paths.py::TestFindAgentBrowser::test_raises_when_not_found. This test attempts to locate a browser binary on the Linux CI runner and times out at 30s. It is unrelated to the changes in this PR — the file is not touched by any of the three commits. The test (4) failure (test_web_search_tool) was caused by our mock dispatcher and has been fixed in the latest commit (f381a34de). A re-run should confirm the fix. The browser timeout may need to be addressed separately (e.g., increasing the timeout or mocking the subprocess call).

**Review by egilewski — APPROVED** [review]
Recommendation: approve

I re-reviewed PR head `f381a34de3e3e165fb4ffecbd3714395522bbde2` against current GitHub main `a72bb03757c0c925c686f9774eefc8dc5a77b329`.

Validation:
- GitHub reports the PR is mergeable, and the current PR-head checks are successful except skipped jobs.
- Replayed the six PR commits onto current `main` in a run-owned worktree; replay applied cleanly.
- `git diff --cached --check`: passed.
- `/home/mac/hermes-agent/.venv/bin/python -B -m compileall -q tools/code_execution_tool.py tests/tools/test_code_execution.py tests/tools/test_code_execution_modes.py tests/tools/test_code_execution_windows_env.py`: passed.
- Direct probe of `SANDBOX_ALLOWED_TOOLS` and generated stubs: `terminal_allowed=False`, `terminal_stub=False`, `web_search_stub=True`.
- Focused prior-failure slice covering the migrated `test_code_execution.py` and `test_code_execution_modes.py` cases: `6 passed`.
- Focused generation and Windows UTF-8 slice: `15 passed, 1 skipped`.
- Broader affected execute-code run, `tests/tools/test_code_execution.py tests/tools/test_code_execution_modes.py tests/tools/test_code_execution_windows_env.py tests/tools/test_execute_code_approval_cluster.py`: `149 passed, 3 skipped`.
- CodeRabbit final clean-pass review against current `main`: no findings.

Finding:
The previously requested terminal-import test migrations are now complete in the reviewed scope. I did not find a remaining blocker.

Signed: GPT-5.5-xhigh in Codex

**Comment by LifeJiggy** [comment]
@teknium1 PTAL thanks

**Review by teknium1 — COMMENTED** [review]
Thanks for carrying the test migration through the earlier CI feedback. The requested allowlist change still has a real target on current main: `tools/code_execution_tool.py:62-70` continues to expose `terminal` to generated sandbox stubs.

### Problems
- Removing the allowlist entry alone leaves terminal-specific guidance unconditional in the execute-code schema at `tools/code_execution_tool.py:1861-1867`. The runtime schema is built from the filtered sandbox set in `model_tools.py:458-463`, so this can advertise `terminal()` after it is no longer importable.
- The updated generation tests need a direct regression for the new contract: current `tests/tools/test_code_execution.py:89-103` is the prior positive coverage, but the PR does not assert that `generate_hermes_tools_module(['terminal'])` now omits `def terminal(`.

### Suggested changes
- Gate or remove terminal-only schema prose and add coverage that the default execute-code schema does not advertise terminal after the allowlist change.
- Add a direct negative generated-stub test for an explicit `terminal` request, while keeping a non-terminal positive stub test.

Automated hermes-sweeper review.

File: tools/code_execution_tool.py [hunk 1]
```diff
@@ -65,7 +65,6 @@
     "write_file",
     "search_files",
     "patch",
-    "terminal",
```
Comment (teknium1) [inline tools/code_execution_tool.py:69]: Please add a direct regression test for this removal: `generate_hermes_tools_module(['terminal'])` should not contain `def terminal(`. The existing excluded-tool test only excludes terminal through a session-specific enabled-tools list, so it would have passed before this allowlist change.

**Comment by LifeJiggy** [comment]
Addressed all reviewer feedback:

1. Removed terminal from SANDBOX_ALLOWED_TOOLS (tools/code_execution_tool.py:62-70) — execute_code runs arbitrary Python; a terminal() call inside that Python lets the model bypass the dangerous-command approval gate. Updated the comment to explain the rationale (references #30882, #33057).

2. Gated terminal-specific schema prose (tools/code_execution_tool.py:1861-1867) — "terminal() is foreground-only" now only appears when terminal is in the enabled sandbox tools. cwd_note references "the terminal tool" instead of terminal() function call. json_parse description no longer references terminal().

3. Added negative regression test (test_code_execution.py):
- test_terminal_excluded_from_sandbox_stubs: generate_hermes_tools_module(["terminal"]) must NOT contain def terminal( — the exact test requested by reviewer
- test_terminal_not_in_sandbox_allowed_tools: asserts terminal is not in SANDBOX_ALLOWED_TOOLS
- test_schema_omits_terminal_prose_when_excluded: schema description must not mention terminal() when excluded
- test_schema_includes_terminal_prose_when_enabled: schema description includes terminal notes when explicitly enabled

4. Updated existing tests to use non-terminal tools (web_search, read_file) as inputs, since terminal is no longer a valid sandbox tool.

Test results: 16/16 pass.

**Comment by egilewski** [comment]
suggesting changes

CI is currently failing on this PR head in `All required checks pass`. The comparable current `main` status is not failing for those check(s), so deeper review should wait for the PR-specific required-check failure to be fixed.

Please fix or rerun the failing check, then push a new head or ask for re-review.

Signed: GPT-5.6-low in Codex

**Comment by LifeJiggy** [comment]
thanks for the review. The CI failures have been fixed.

Root cause: The first commit removed terminal from SANDBOX_ALLOWED_TOOLS (the security fix), but several tests in test_code_execution.py and test_code_execution_modes.py still used from hermes_tools import terminal inside sandbox scripts. Since the hermes_tools stubs are only generated for tools in SANDBOX_ALLOWED_TOOLS, these imports fail with ImportError on Linux CI (Windows skips them via @skipIf).

Fix (commit ad0e5f80c): Replaced terminal with web_search (which IS in SANDBOX_ALLOWED_TOOLS) in 9 tests across both files:
- test_single_tool_call, test_multi_tool_chain, test_concurrent_tool_calls_match_responses — basic tool call tests
- test_none_enabled_tools_uses_all, test_empty_enabled_tools_uses_all, test_nonoverlapping_tools_fallback — edge case tests
- test_project_mode_can_still_import_hermes_tools, test_strict_mode_can_still_import_hermes_tools — import regression tests

The test_excluded_tool_returns_error test still uses terminal intentionally — it tests that calling an excluded tool produces an error, which is still the correct behavior (import fails).

The 4 TestRpcTokenAuthorization failures are pre-existing Windows-only (socket.AF_UNIX not available). Same failures exist on origin/main.
